### PR TITLE
Quickcache unicode/string vary on

### PR DESCRIPTION
**Pull request by snopoke**
@dannyroberts 

@TylerSheffels review buddy

### CODE REVIEW

File: corehq/util/quickcache.py [hunk 1]
```diff
@@ -87,14 +95,20 @@ def _hash(value, length=32):
...
+            # the user to explicitly request the different behaviour.
+            if isinstance(value, unicode):
+                encoded = value.encode('utf-8')
             else:
-                # for long or non-alphanumeric text
-                return 't' + self._hash(value)
+                try:
+                    encoded = value.decode('utf-8').encode('utf-8')
+                except UnicodeDecodeError:
+                    self.encoding_assert(False, 'Non-utf8 encoded string used as cache vary on')
```
Comment (snopoke) [inline corehq/util/quickcache.py]: @dannyroberts not sure if this is necessary, what do you think?

File: corehq/util/quickcache.py [hunk 2]
```diff
@@ -65,6 +67,12 @@ def __init__(self, fn, vary_on, cache):
                         'no such argument'.format(arg, self.fn.__name__)
                     )
 
+        self.encoding_assert = soft_assert(
+            to=['{}@{}'.format('skelly', 'dimagi.com')],
```
Comment (dannyroberts) [inline corehq/util/quickcache.py]: nice

File: corehq/util/quickcache.py (hunk 1 above)
Comment (dannyroberts) [inline corehq/util/quickcache.py]: I think if the purpose of this is to disallow passing in binary data (for example), but you wanna make for sure for sure that it doesn't ever happen, then this is appropriate. There'll have to be another culling step where you remove this and replace it with an assert in a month or whenever you think of it

Comment (snopoke) [inline corehq/util/quickcache.py]: I think what I want is to see if this does happen and whether it's likely
to result in problems down the road.

**Comment by dannyroberts** [comment]
Looks like `test_vary_on_func` is failing because the cache key format changed (?)

**Comment by dannyroberts** [comment]
May have been too brittle a test that I wrote

**Comment by dannyroberts** [comment]
:+1:

**Pull request merged**